nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs

```diff
@@ -333,6 +276,139 @@ impl<'a> BuilderExternalNetworking<'a> {
     }
 }
 
+// Helper to validate that the system hasn't gone off the rails. There should
+// never be any external networking resources in the planning input (which is
+// derived from the contents of CRDB) that we don't know about from the parent
+// blueprint.
+//
+// There may still be database records corresponding to _expunged_ zones, but
+// that's okay: it just means we haven't yet realized a blueprint where those
+// zones are expunged. And those should should still be in the blueprint (not
+// pruned) until their database records are cleaned up.
+//
+// It's also possible that there may be networking records in the database
+// assigned to zones that have been expunged, and our parent blueprint uses
+// those same records for new zones. This is also fine and expected, and is a
+// similar case to the previous paragraph: a zone with networking resources was
+// expunged, the database doesn't realize it yet, but can still move forward and
+// make planning decisions that reuse those resources for new zones.
+fn ensure_input_records_appear_in_parent_blueprint(
+    parent_blueprint: &Blueprint,
+    input: &PlanningInput,
+) -> anyhow::Result<()> {
+    let mut all_macs: HashSet<MacAddr> = HashSet::new();
+    let mut all_nexus_v4_ips: HashSet<Ipv4Addr> = HashSet::new();
+    let mut all_nexus_v6_ips: HashSet<Ipv6Addr> = HashSet::new();
+    let mut all_boundary_ntp_v4_ips: HashSet<Ipv4Addr> = HashSet::new();
+    let mut all_boundary_ntp_v6_ips: HashSet<Ipv6Addr> = HashSet::new();
+    let mut all_external_ips: HashSet<OmicronZoneExternalIp> = HashSet::new();
+
+    // Unlike the construction of the external IP allocator and existing IPs
+    // constructed above in `BuilderExternalNetworking::new()`, we do not
+    // check for duplicates here: we could very well see reuse of IPs
+    // between expunged zones or between expunged -> running zones.
+    for (_, z) in parent_blueprint.all_omicron_zones(BlueprintZoneFilter::All) {
+        let zone_type = &z.zone_type;
+        match zone_type {
+            BlueprintZoneType::BoundaryNtp(ntp) => match ntp.nic.ip {
+                IpAddr::V4(ip) => {
```

Comment:
Just a suggestion: could we make the sets contain `IpAddr`?  That'd consolidate it into fewer sets and eliminate the matching.

edit: I see that would make the code below a little more complicated.  I'll leave it to you to decide if it's any better on net.

Reply:
It's an improvement; the code below just needs some cheap `.into()`s ([9ae92dc](https://github.com/oxidecomputer/omicron/pull/6599/commits/9ae92dc9bbaeeadf149f23c26f3d640311c514b0)). Thanks!